### DISCUSSION TITLE
How to run NiceGUI behind and IIS proxy?

### DISCUSSION THREAD

**Discussion by vindevoy** [zauberzeug/nicegui#1600] [Q&A] [answered]
### Question

At work, we use IIS from Microsoft (yes, unfortunately) to run the web servers.  Sometimes because we can use the integrated Windows security, but that is not the case here.  I need simple proxying of port 8000 to 80.

To do so, I managed to create an inbound rule that does this.
- Website: default website
- Pattern match: (.*)
- URL: http://{SERVER_NAME}:8000/{R:1}

This works fine when I make this a "redirect" url.  That works, but I would like to hide the port 8000 and eventually even close this port on the firewall.

However, when I make this a "rewrite" URL, the port is hidden and I see the site on the default http port.  However, in that case, the tree component and the buttons that open a new page in a single page application stop to work.

I don't find that surprising because I know there's a backtrip to the server.  But I can't make a outbound rule for rewriting.  As soon as I set a rule, I think it's IIS that gives me an internal server error.  I've tried it via the "inbound and outbound" rules also, but then also the outbound rule results in a 500 error.

Has anybody any idea how to run a nicegui app with tree and buttons behind IIS ?

**Comment by rodja** [comment]
I've no experience with ISS, but you need to configure the proxy in a way to allow websocket connections. For Nginx it's done like this: https://github.com/zauberzeug/nicegui/blob/97261a6a1890476d91d4bbdb2125196c8e46a128/examples/nginx_subpath/nginx.conf#L46-L50

**Reply by vindevoy** [reply]
Yes, I would prefer also an NGinx, but unfortunately ...  I have to set this IIS stuff via the IIS manager and I have not enough knowledge of this thing.  I still wonder why their own suggestions don't work for the reverse proxy.  You see, you fill out the proxy part and they propose the reverse proxy part.  Which does not work ... Imagine !  It gives an internal server error.

**Comment by vindevoy** [comment]
I have successfully proxy forwarded towards the nicegui application.  However, I think I miss the proxy reverse in a correct way.  I am correct when I say that on a button click event, or on a tree select event, there is communication with the backend, right ?  I suppose that's where it goes wrong.  Although, when I add a ui.notify for debugging, I see the URL is just a short path without the hostname and so.  Can't get my head around that part.

**Comment by rodja** [comment]
> I am correct when I say that on a button click event, or on a tree select event, there is communication with the backend, right ?

Yes. If button click works you have communication with the backend.

> Although, when I add a ui.notify for debugging, I see the URL is just a short path without the hostname and so. Can't get my head around that part.

I do not understand. Are you using `ui.notify` to display the url?

**Reply by vindevoy** [reply]
Yes, I use a ui.notify as debugging.  To see if I for instance could see the hostname in the URL.  My app is running on server:80, my button points to server:80/static/index.html, but if the URL would say server:8000/static/index.html, I could imagine that the button cannot connect to the backend.  Which is not the case because I have 8000 open for the moment, but still, somewhere some security that prevents connecting if host and port are not the same.

But the URL was /static/index.html.  So relative.  I don't see why it cannot communicate with the backend then.

**Reply by vindevoy** [reply]
On my button click event, I use a lambda : my_function(params).  In my_function, I start with a ui.notify as debugging, then the rest of my code.

**Reply by rodja** [reply]
Sorry. I cant follow. I don't think it's fruitful to test the websocket communication by showing url paths. Let's establish the facts one by one: Does a button like `ui.button('check', on_click=lambda: ui.notify('bam!'))` work when accessing through the IIS rewrite? Eg. do you see "bam!" notification if you press it?

**Reply by vindevoy** [reply]
I confirm that this is not working.  I do not see the notification.

I created a completely new application, 3 lines, your button example.  The sample application works when I start it using:

`python main.py`

and open the browser on port 8080.  When I click the button, I receive the notification.


I set up IIS like the screenshot below.  

![image](https://github.com/zauberzeug/nicegui/assets/26721992/b82ff621-704e-4a1e-98a3-881b0be9d885)

I can reach the application on http://<server_name> without problems.  I see the button.  When I click the button, there is no notification.


I  have not setup a reverse rule yet.  As I can't figure out how to give the outbound rule.  When I type the exact same thing in a "proxy rule", IIS saves the outbound rule and then displays a "500 internal server error" when I access the page.

I must setup the inbound rule like this, as I'm using a NiceGui SPA application where I have multiple URL's I need to forward.  This rule works for showing each page in my application, but the buttons don't work.  Nor the tree.

**Reply by rodja** [reply]
Thanks for the information. Unfortunately I'm completely lost about the features and settings of IIS. Can anyone help out?

**Reply by foinus** [reply, marked as answer]
hi I managed to get this to work, I set up my rewrite rule like yours, the only difference is I put Onboarding/(.*) in my pattern and root_path of my ui.run() (I do not think this has anything to do with your problem) 
the thing that made it work for me was to set a system variable HTTP_SEC_WEBSOCKET_EXTENSIONS set to 0
like in this [article](https://www.oxygenxml.com/doc/versions/24.1.0/ug-waCustom/topics/WA-websocket.html)

future more u need to Enable reverse proxy settings in IIS like in this [article](https://ded9.com/configuring-iis-reverse-proxy-server-windows-server/#:~:text=Step%203.%20Enable%20reverse%20proxy%20settings%20in%20IIS)

![image](https://github.com/user-attachments/assets/99fd7711-649a-4132-bfe4-1af8b6af59b6)
